### PR TITLE
chore(performance_tests): added hierarchy keyring and caching cmm tests

### PR DESCRIPTION
**Pull request by RitvikKapila**
*Issue #, if available:*

*Description of changes:*
Added HKeyring and CachingCMM tests

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: performance_tests/src/aws_encryption_sdk_performance_tests/master_key_providers/caching_cmm.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Performance tests for the Caching Cryptographic Materials Manager (CMM)."""
```
Comment (lucasmcdonald3) [inline performance_tests/src/aws_encryption_sdk_performance_tests/master_key_providers/caching_cmm.py:3]: We need to note that this is the caching CMM with the KMS MKP.

Ideally we might have considered this in our performance test's input dimensions:
```
key_provider: [kms mkp, raw rsa mkp, raw kms mkp]
cmm: [default, caching]
```

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Performance tests for the hierarchy keyring."""
+
+# noqa pylint: disable=wrong-import-order
```
Comment (lucasmcdonald3) [inline performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:5]: Do we need this?

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    )
+
+    # Create the Hierarchical Keyring.
+    mat_prov: AwsCryptographicMaterialProviders = AwsCryptographicMaterialProviders(
+        config=MaterialProvidersConfig()
+    )
+
+    keyring_input: CreateAwsKmsHierarchicalKeyringInput = CreateAwsKmsHierarchicalKeyringInput(
+        key_store=keystore,
+        branch_key_id_supplier=branch_key_id_supplier,
```
Comment (lucasmcdonald3) [inline performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:115]: Let's just one one branch key ID. A supplier is overkill for a performance test
```suggestion
        branch_key_id=...
```

File: performance_tests/src/aws_encryption_sdk_performance_tests/master_key_providers/caching_cmm.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    :param kms_key_id: Amazon Resource Name (ARN) of the KMS customer master key
+    :type kms_key_id: str
+    :param max_age_in_cache: Maximum time in seconds that a cached entry can be used
+    :type max_age_in_cache: float
+    :param cache_capacity: Maximum number of entries to retain in cache at once
+    :type cache_capacity: int
+    """
+    # Security thresholds
+    #   Max messages (or max bytes per) data key are optional
+    max_entry_messages = 100
```
Comment (lucasmcdonald3) [inline performance_tests/src/aws_encryption_sdk_performance_tests/master_key_providers/caching_cmm.py:25]:
```suggestion
    max_messages_encrypted = 100
```
I'm not sure what `max_entry_messages` is supposed to mean. Just name it the variable name you're assigning it to

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    DefaultCache,
+    GetBranchKeyIdInput,
+    GetBranchKeyIdOutput,
+)
+from aws_cryptographic_materialproviders.mpl.references import IBranchKeyIdSupplier, IKeyring
+
+from ..utils.util import PerfTestUtils
+
+
+class ExampleBranchKeyIdSupplier(IBranchKeyIdSupplier):
```
Comment (lucasmcdonald3) [inline performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:27]: Remove this, just one one branch key ID

File: performance_tests/src/aws_encryption_sdk_performance_tests/master_key_providers/caching_cmm.py (hunk 1 above)
Comment (RitvikKapila) [reply performance_tests/src/aws_encryption_sdk_performance_tests/master_key_providers/caching_cmm.py:3]: Updated in the doc string for the file

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py (hunk 2 above)
Comment (RitvikKapila) [reply performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:5]: pylint was throwing some weird errors and I added a bunch of these to debug and missed it while removing. Removed now. Thanks

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py (hunk 3 above)
Comment (RitvikKapila) [reply performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:115]: Done

File: performance_tests/src/aws_encryption_sdk_performance_tests/master_key_providers/caching_cmm.py (hunk 4 above)
Comment (RitvikKapila) [reply performance_tests/src/aws_encryption_sdk_performance_tests/master_key_providers/caching_cmm.py:25]: I saw this in the example: https://github.com/aws/aws-encryption-sdk-python/blob/ebbc26b1fb83d7e0bcba66c70012730350b6c60d/examples/src/legacy/data_key_caching_basic.py#L20

Updated now,

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py (hunk 5 above)
Comment (RitvikKapila) [reply performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:27]: Done

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+from aws_cryptographic_materialproviders.mpl.references import IKeyring
+
+from ..utils.util import PerfTestUtils
+
+
+def create_keyring(
+    key_store_table_name: str,
+    logical_key_store_name: str,
+    kms_key_id: str
+):
```
Comment (lucasmcdonald3) [inline performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:26]:
```suggestion
def create_keyring(
    key_store_table_name: str,
    logical_key_store_name: str,
    kms_key_id: str,
    branch_key_id: str = PerfTestUtils.DEFAULT_BRANCH_KEY_ID
):
```

More configurable

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            logical_key_store_name=logical_key_store_name,
+            kms_client=kms_client,
+            kms_configuration=KMSConfigurationKmsKeyArn(
+                value=kms_key_id
+            ),
+        )
+    )
+
+    # Call CreateKey to create a new branch key.
+    branch_key_id: str = PerfTestUtils.DEFAULT_BRANCH_KEY_ID
```
Comment (lucasmcdonald3) [inline performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:60]:
```suggestion
```

File: performance_tests/test/keyrings/test_hierarchy_keyring.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""This is a performance test for creating the hierarchy keyring."""
+
+import os
+# noqa pylint: disable=wrong-import-order
```
Comment (lucasmcdonald3) [inline performance_tests/test/keyrings/test_hierarchy_keyring.py:6]: Needed?

**Review by lucasmcdonald3 — APPROVED** [review]

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py (hunk 7 above)
Comment (RitvikKapila) [reply performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:60]: done

File: performance_tests/test/keyrings/test_hierarchy_keyring.py (hunk 8 above)
Comment (RitvikKapila) [reply performance_tests/test/keyrings/test_hierarchy_keyring.py:6]: removed. good catch. thanks

File: performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py (hunk 6 above)
Comment (RitvikKapila) [reply performance_tests/src/aws_encryption_sdk_performance_tests/keyrings/hierarchy_keyring.py:26]: Yes this is better. Thanks

**Pull request merged**